### ISSUE TITLE
PBTS: support both PBTS and legacy BFT Time

### ISSUE THREAD

**Issue by glnro** [cometbft/cometbft#2063]
In order to ensure backwards compatibility, PBTS should be enabled via consensus params.

In #2018 `MedianTime` was removed per the original work in tendermint/tendermint#7382. This does not consider that PBTS can be adopted in existing chains, with blocks produced and that need to be verified using the BFT Time algorithm. 

### Specification

- Addressed by #2415 (depends on 2376 below)
  - [x] Specify the transition from BFT Time to PBTS via consensus parameter
     - Add it to `spec/consensus/time.md`
  - [x] Update documentation of `FeatureParams`, remove references to `ABCIParams`

### Documentation

- Addressed by #2376
  - [x] Update ADR with the adopted solution: #2223
  - [x] Update the [PBTS documentation](https://github.com/cometbft/cometbft/blob/feature/pbts/docs/explanation/core/proposer-based-timestamps.md) accordingly
  - [x] Add to the [PBTS documentation](https://github.com/cometbft/cometbft/blob/feature/pbts/docs/explanation/core/proposer-based-timestamps.md) the workaround for transitioning to PBTS proposed in the ADR (see this [comment](https://github.com/cometbft/cometbft/pull/2223#discussion_r1481259421)).

### Implementation

- "Rescue" code removed by cherry-picks:
  - [x] #2203
    - #2205
  - [x] #2204
    - #2205
  - [x] #2397 
  - Adapt calls to the construction `types.NewProposal` to accept a timestamp (see [comment](https://github.com/cometbft/cometbft/pull/2018#discussion_r1458000432))
      - This is not needed, as in if we didn't update them yet, compilation errors and test failing would be observed
- Introduce and implement consensus param to enable PBTS
  - [x] #2197
    - #2231
    - #2325
  - [x] #2326
  - [x] #2438

### Testing

- [x] #2324
   - https://github.com/cometbft/cometbft/pull/2329
- [x] #2393
   - https://github.com/cometbft/cometbft/pull/2404 (depends on 2339 above)
- [x] #2227

**Comment by cason** [comment]
Some relevant a priori questions:

1. Should we still support BFT Time?
2. Should we allow chains to stick with BFT Time, i.e., to not adopt PBTS from v1.0?

**Comment by sergio-mena** [comment]
Summary of decisions made during the meeting today (participants @glnro, @lasarojc, @cason , @sergio-mena):
* Calls to function `MedianTime` could just be removed from the `blocksync` reactor and this wouldn't affect the integrity/security of the blockchain, even in BFT-time mode
  * So, the cherry-picked code from 2022 would be functionally complete, no new functionally would be needed, if we were to activate PBTS right at the upgrade height.
* HOWEVER, because of product concerns, we don't want PBTS to be on the critical path when upgrading to `v1.0`
  * We prefer a method were the users have the possibility to delay PBTS activaton, thus detaching it from the upgrade to `v1.0`
* THEREFORE, we keep the idea of adding a consensus param containing the height where PBTS comes into force
  * The way this param works is analogous to `VoteExtensionEnableHeight`, but the height $H$ in question is _not_ different from $H+1$ and later (unlike what happens with `VoteExtensionEnableHeight`)
  * This new param will allow us to keep function `MedianTime`, which will still be called when still on BFT-time, both from consensus and from blockchain

Tasks proposed (to be updated in tracking issue):

* Update spec (new section explaining the consensus param?)
  * [ ] #2020
* Implementation: follow tasks already laid out in this issue's description
  * Basically "rescue" code  that was removed by the cherry-picks
* Add logic to use the new consensus param to select the rescued code, or not (depending on whether PBTS is already activated)
  * "rescued" : BFT-time related code that was removed by the cherry-picks
* Add a new UT (or modify some existing one) in `pbts_test.go` to test the activation functionality
* Add infra in `e2e` (and generator) to acceptance-test this

**Comment by sergio-mena** [comment]
About `e2e` infra: don't forget to implement a manifest option "clock_skew", which will also be used by the manifest generator

**Comment by cason** [comment]
> Calls to function `MedianTime` could just be removed from the blocksync reactor and this wouldn't affect the integrity/security of the blockchain, even in BFT-time mode 

To make it clear. If we are running a chain with PBTS and validate _old_ blocks, produced using BFT Time, we don't have problems. For producing _new_ blocks in a BFT Time-enabled chain, we need the referred check.

**Comment by sergio-mena** [comment]
> For producing new blocks

Yes, my way of saying _old blocks_ in the text was "from the blocksync reactor" (new blocks are created in consensus).
So I think we're on the same page here.
